state/upgrades.go

```diff
@@ -1430,3 +1430,45 @@ func CopyMongoSpaceToHASpaceConfig(st *State) error {
 	_, err = settings.Write()
 	return errors.Annotate(err, "writing controller info")
 }
+
+// CreateMissingApplicationConfig ensures that all models have an application config in the db.
+func CreateMissingApplicationConfig(st *State) error {
+	settingsColl, settingsCloser := st.db().GetRawCollection(settingsC)
+	defer settingsCloser()
+
+	var applicationConfigIDs []struct {
+		ID string `bson:"_id"`
+	}
+	settingsColl.Find(nil).All(&applicationConfigIDs)
+
+	allIDs := set.NewStrings()
+	for _, id := range applicationConfigIDs {
+		allIDs.Add(id.ID)
+	}
+
+	appsColl, appsCloser := st.db().GetRawCollection(applicationsC)
+	defer appsCloser()
+
+	var applicationIDs []struct {
+		Name      string `bson:"name"`
+		ModelUUID string `bson:"model-uuid"`
+	}
+	appsColl.Find(nil).All(&applicationIDs)
+
+	var newAppConfigs []txn.Op
+	emptySettings := make(map[string]interface{})
+	// var emptySettings map[string]interface{}
+	for _, app := range applicationIDs {
+		appConfID := fmt.Sprintf("%s:a#%s#application", app.ModelUUID, app.Name)
+		if !allIDs.Contains(appConfID) {
+			newOp := createSettingsOp(settingsC, appConfID, emptySettings)
+			newOp.Insert.(*settingsDoc).ModelUUID = app.ModelUUID
+			newAppConfigs = append(newAppConfigs, newOp)
+		}
+	}
+	err := st.db().RunRawTransaction(newAppConfigs)
```

Comment:
Do we have to worry about this ever being executed part-way and then re-run later?
I'm a bit concerned that the whole loop assumes nothing is changing in the database, and will fail if some other thing starts inserting no-op settings entries.
(It seems like the whole thing should be wrapped in the retry loop that re-evaluates what items need to be created.)
I do believe upgrade steps are supposed to be run sequentially and only by one controller, and if the entire upgrade is restarted, it will handle it, so it isn't critical.
Just something we ran into with things like the leases collection.

Reply:
To clarify, with thanks from @babbageclunk, .Run doesn't work across models so we use runRawTransaction. The thinking was that anything that creates no-op settings will cause the transaction to fail and thus the upgrade step would be re-run (as long as the thing creating the no-ops isn't doing it continuously we get to work out ok).